setup/util/util.go

```diff
@@ -0,0 +1,154 @@
+		return errors.New("value must be numeric")
+	}
+	if ansNum < 10 && ansNum != -1 {
+		return errors.New("value must be either -1 to disable or 10 or greater")
+	}
+	return nil
+}
+
+var expectedMd5s = map[string]string{
+	"explain":              "7a0a1784d170975d8538d3b8b38c3fad",
+	"get_stat_replication": "4cc3b87c1f030eeb680d4f0bd2732836",
+}
+
+func ValidateHelperFunction(fn string, runner *query.Runner) (bool, error) {
+	expected, ok := expectedMd5s[fn]
+	if !ok {
+		return false, fmt.Errorf("unrecognized helper function %s", fn)
+	}
+	row, err := runner.QueryRow(
+		fmt.Sprintf(
+			`SELECT md5(btrim(prosrc, E' \\n\\r\\t'))
+FROM pg_proc INNER JOIN pg_user ON (pg_proc.proowner = pg_user.usesysid)
+WHERE proname = %s
+	AND pronamespace::regnamespace::text = 'pganalyze'
+	AND prosecdef
+  AND pg_user.usesuper`,
+			pq.QuoteLiteral(fn),
+		),
+	)
+	if err == query.ErrNoRows {
+		return false, nil
+	} else if err != nil {
+		return false, err
+	}
+	actual := row.GetString(0)
+	return actual == expected, nil
+}
+
+func GetPendingSharedPreloadLibraries(runner *query.Runner) (string, error) {
+	// When shared_preload_libraries is updated, since the setting requires a restart for the
+	// changes to take effect, the new value is not reflected with SHOW or current_setting().
+	// To make sure we don't clobber any pending changes (including our own, if adding both
+	// pg_stat_statements *and* auto_explain), we need to read the configured-but-not-yet-applied
+	// value from the config file (there does not appear to be a better way to do this)
+
+	// N.B.: we project name here even though we don't explicitly need it,
+	// because a valid (and in fact, common) value for shared_preload_libraries
+	// is the empty string, and because our query mechanism depends on CSV, and
+	// because of https://github.com/golang/go/issues/39119 , that value cannot
+	// be parsed correctly by Go's encoding/csv if that's the only value in the
+	// CSV file output.
+
+	// N.B.: note also that checking sourcefile/sourceline here is a heuristic: these
+	// describe where the *current* value comes from (not the pending value), but this
+	// is our best guess.
+	row, err := runner.QueryRow(`
+SELECT
+  name,
+  CASE
+    WHEN pending_restart THEN
+      left(
+        right(
+          regexp_replace(
+            (SELECT line FROM
+              (SELECT row_number() OVER () AS line_no, line FROM
+                regexp_split_to_table(
+                  pg_read_file(
+                    COALESCE(sourcefile, 'postgresql.auto.conf')
+                  ), '\s*$\s*', 'm'
+                ) AS lines(line)
+              ) AS numbered_lines(line_no, line)
+             WHERE
+               CASE WHEN sourceline IS NULL THEN line LIKE name || ' = %' ELSE line_no = sourceline END
+            ),
+            name || ' = ', ''
+          ),
+      -1),
+    -1)
+    ELSE current_setting(name)
+  END AS pending_value
+FROM
+  pg_settings
+WHERE
+  name = 'shared_preload_libraries'`)
```

Comment:
TODO: @lfittl to review this once more

Reply:
Reviewing this, I can follow the logic, however I think there is a bug here:

It seems that we skip over empty lines in the inner query:

```
SELECT row_number() OVER () AS line_no, line FROM
                regexp_split_to_table(
                  pg_read_file(
                    COALESCE(sourcefile, 'postgresql.auto.conf')
                  ), '\s*$\s*', 'm'
                ) AS lines(line)
```

And this then throws off the `line_no = sourceline` comparison in the outer SELECT. I think we need to preserve empty lines there, so the line number comparison works.

Reply:
It seems there is another problem here as well, related to how Postgres determines line numbers, or rather when it sets the `sourceline` value. It seems we can't trust what we're getting from `pg_settings`.

postgresql.auto.conf:

```
# Do not edit this file manually!
# It will be overwritten by the ALTER SYSTEM command.
log_line_prefix = '%m [%p] %q[user=%u,db=%d,app=%a] '
shared_preload_libraries = 'pg_stat_statements, auto_explain'
```

```
postgres=# SELECT                  
  name, sourcefile, sourceline FROM
  pg_settings
WHERE
  name = 'log_line_prefix';
      name       |                    sourcefile                    | sourceline 
-----------------+--------------------------------------------------+------------
 log_line_prefix | /var/lib/postgresql/13/main/postgresql.auto.conf |          3
(1 row)

postgres=# SELECT                  
  name, sourcefile, sourceline FROM
  pg_settings
WHERE
  name = 'shared_preload_libraries';
           name           |                    sourcefile                    | sourceline 
--------------------------+--------------------------------------------------+------------
 shared_preload_libraries | /var/lib/postgresql/13/main/postgresql.auto.conf |          3
(1 row)
```

The `shared_preload_libraries` value here is of course wrong, which then leads to us getting the wrong value, and this query not working. Interestingly, after a restart this was fixed. It seems this may be related to subsequent `ALTER SYSTEM` commands not updating the `sourceline` correctly? This was with the regular setup flow, which first sets `shared_preload_libraries = 'pg_stat_statements'` (which would make it line 3), then sets `log_line_prefix` (which presumably adds it first in line 4), and then again `shared_preload_libraries = 'auto_explain, pg_stat_statements'`, which causes the order to change, but apparently doesn't update the `sourceline` value correctly.

If we only match based on the line pattern, this works as expected:

```
SELECT
  name,
  CASE
    WHEN pending_restart THEN
      left(
        right(
          regexp_replace(
            (SELECT line FROM
              (SELECT row_number() OVER () AS line_no, line FROM
                regexp_split_to_table(
                  pg_read_file(
                    COALESCE(sourcefile, 'postgresql.auto.conf')
                  ), '\s*$\s*', 'm'
                ) AS lines(line)
              ) AS numbered_lines(line_no, line)
             WHERE
               line LIKE name || ' = %'                                                                
            ),
            name || ' = ', ''
          ),
      -1),
    -1)
    ELSE current_setting(name)
  END AS pending_value
FROM
  pg_settings
WHERE
  name = 'shared_preload_libraries';

           name           |          pending_value           
--------------------------+----------------------------------
 shared_preload_libraries | pg_stat_statements, auto_explain
(1 row)
```

Reply:
Yeah, you're right--I think the issue is that `pg_settings` has the `sourcefile` and `sourceline` of the *current* setting. Essentially it's metadata about the snapshot of the config file from which the setting as applied was loaded (and when we read it, the contents may be different than the snapshot). We `pg_reload_conf` after every change so this is not an issue *except* in the very important case of `shared_preload_libraries` post change but before restart.

I think the solution is as you're suggesting: skip the line number and just match on the name. I was using that as a fallback because I thought the line number method was higher fidelity, but I did not account for the subtle semantics here.

Reply:
Thinking about this some more, even `sourcefile` can be bogus here, right? Consider the following:

1. I have an existing `shared_preload_libraries = some_module` manually set in `postgresql.conf`
2. I run the `pg_stat_statements` step which correctly finds that based on `sourcefile`, adds `pg_stat_statements`, and saves the resulting `some_module, pg_stat_statements` to `postgresql.auto.conf` using `ALTER SYSTEM`
3. I run the `auto_explain` step, which I think at this point (pre-restart), will still incorrectly see `postgresql.conf` as the `sourcefile` (because it's the source of the current pre-restart value), grab the out-of-date value, and incorrectly save the new value of `some_module, auto_explain`, losing `pg_stat_statements`.

Maybe given this, the right thing to do here is if the key is set in postgresql.auto.conf, always use that value, and only fall back to `sourcefile` when it is not present there?